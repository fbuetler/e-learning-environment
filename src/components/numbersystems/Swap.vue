<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <div>Stelle die gleiche Summe mit weniger Elementen dar</div>
    <ItemGroup :items="generatedItems" :itemConfig="items(type)" />
    <hr />
    <div>LÃ¶sung:</div>
    <ItemDropzone
      :items="selectedItems"
      :itemConfig="items(type)"
      @dropped="addItem()"
    />
    <div
      class="interaction-container flex-item flex-row flex-center flex-stretch"
    >
      <ItemSelection
        :selected="selected"
        :items="items(type)"
        @selected="selected = $event"
      />
      <Undo @undo-operation="undo()" />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import NumbersystemsMixin, {
  numbersystemType,
} from "@/components/numbersystems/NumbersystemsMixin.vue";
import Undo from "@/components/Undo.vue";
import Difficulty from "@/components/Difficulty.vue";
import ItemSelection from "@/components/ItemSelection.vue";
import ItemDropzone from "@/components/numbersystems/ItemDropzone.vue";
import ItemGroup from "@/components/numbersystems/ItemGroup.vue";

@Component<Swap>({
  components: {
    Undo,
    Difficulty,
    ItemSelection,
    ItemDropzone,
    ItemGroup,
  },
})
export default class Swap extends Mixins(GameMixin, NumbersystemsMixin)
  implements GameInterface {
  @Prop({ required: true })
  args!: { numbersystemType: numbersystemType };
  type = this.args.numbersystemType;

  selected: number = null;
  selectedItems: Array<number> = null;
  generatedItems: Array<number> = null;
  animationSteps: Array<string> = null;

  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  isInitialized(): boolean {
    return this.generatedItems !== null;
  }

  start() {
    this.selected = null;
    this.selectedItems = new Array<number>(this.items(this.type).length).fill(
      0
    );
    do {
      this.generatedItems = this.generateItems(this.type);
    } while (
      this.sumItems(this.type, this.generatedItems) >= this.limit(this.type) ||
      this.sum(this.generatedItems) ===
        this.sum(
          this.calcMinimalAmount(
            this.type,
            this.sumItems(this.type, this.generatedItems)
          )
        )
    );
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    return (
      this.sumItems(this.type, this.selectedItems) ===
        this.sumItems(this.type, this.generatedItems) &&
      this.sum(this.selectedItems) < this.sum(this.generatedItems)
    );
  }

  sum(arr: number[]): number {
    return arr.reduce((sum, el) => (sum += el));
  }

  addItem() {
    if (this.selected === null) {
      return;
    }
    Vue.set(
      this.selectedItems,
      this.selected,
      this.selectedItems[this.selected] + 1
    );
    this.selected = null;
  }

  undo() {
    for (let i = 0; i < this.selectedItems.length; i++) {
      Vue.set(this.selectedItems, i, 0);
    }
  }

  getAnimationSteps(): Array<string> {
    const correctSum = this.sumItems(this.type, this.generatedItems);
    const wrongSum = Math.ceil(Math.random() * this.limit(this.type));

    return this.mapNumberToActions(wrongSum, this.type)
      .concat(["button-menu-check", "undo"])
      .concat(this.mapNumberToActions(correctSum, this.type))
      .concat(["button-menu-check", "button-menu-next"]);
  }
}
</script>

<style scoped></style>
