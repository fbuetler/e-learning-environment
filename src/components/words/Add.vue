<template>
  <div @dragend="selected = null">
    <slot name="animation" :animationSteps="animationSteps" />
    <div>
      Versuch ein neues Wort zu bilden, indem du einen Buchstaben hinzuf√ºgst.
    </div>
    <div
      id="word-container"
      ref="word-container"
      class="word-container flex-item flex-col flex-center flex-flex"
    >
      <div class="flex-item flex-row flex-center">
        <div
          v-for="element in word"
          :id="`word-char-${element.id}`"
          :ref="`word-char-${element.id}`"
          :key="element.id"
          :class="{
            locked: element.locked,
          }"
          class="word-char card"
        >
          {{ element.char }}
        </div>
        <div :ref="`word-char-${word.length}`"></div>
      </div>
      <div v-if="!charAdded" class="svg-container">
        <svg class="svg-item">
          <defs>
            <marker
              id="arrowhead"
              viewBox="0 0 10 10"
              refX="1"
              refY="5"
              markerWidth="6"
              markerHeight="6"
              orient="auto"
            >
              <path d="M 0 5 L 10 0 L 10 10 z" />
            </marker>
          </defs>
          <line
            v-for="id in arrows"
            :id="`arrow-${id}`"
            :ref="`arrow-${id}`"
            :key="`arrow-${id}`"
            stroke="black"
            stroke-width="3"
            fill="transparent"
            marker-start="url(#arrowhead)"
          />
          <rect
            v-for="id in arrows"
            :id="`rect-around-arrow-${id}`"
            :ref="`rect-around-arrow-${id}`"
            :key="`rect-around-arrow-${id}`"
            fill="transparent"
            @click="addChar(id)"
            @dragenter.prevent
            @dragover.prevent
            @drop.stop.prevent="addChar(id)"
          />
        </svg>
      </div>
    </div>
    <hr />
    <div
      class="interaction-container flex-item flex-row flex-center flex-stretch flex-wrap"
    >
      <ItemSelection
        :selected="selected"
        :items="items"
        @selected="selected = $event"
      />
      <Undo @undo-operation="undo()" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import ItemSelection, { item } from "@/components/shared/ItemSelection.vue";
import Undo from "@/components/shared/Undo.vue";
import {
  LoadWords,
  wordElement,
  alphabet,
  items,
  findCorrectAndWrongSolutions,
} from "@/components/words/Words";

@Component<Add>({
  components: {
    ItemSelection,
    Undo,
  },
})
export default class Add extends Mixins(GameMixin) implements GameInterface {
  word: wordElement[] = null;
  similarWords: string[] = null;
  selected: number = null;
  charAdded = false;

  animationSteps: Array<string> = null;

  created() {
    window.addEventListener("resize", this.drawArrows);
  }

  destroyed() {
    window.removeEventListener("resize", this.drawArrows);
  }

  mounted() {
    this.drawArrows();
  }

  updated() {
    this.drawArrows();
  }

  isInitialized(): boolean {
    return this.word !== null;
  }

  start() {
    [this.word, this.similarWords] = LoadWords("add", 1);
    this.word.forEach((el) => (el.locked = true));
    this.selected = null;
    this.charAdded = false;
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect() {
    return this.similarWords.includes(this.word.map((el) => el.char).join(""));
  }

  drawArrows() {
    if (this.charAdded) {
      return;
    }
    this.arrows.forEach((id) => {
      const container = this.$refs["word-container"] as HTMLElement;
      let char: HTMLElement;
      if (id === this.word.length) {
        char = this.$refs[`word-char-${id}`] as HTMLElement;
      } else {
        // When ref is used together with v-for, the ref you get will be an array
        char = this.$refs[`word-char-${id}`][0] as HTMLElement;
      }
      const arrow = this.$refs[`arrow-${id}`][0] as SVGElement;
      const rect = this.$refs[`rect-around-arrow-${id}`][0] as SVGElement;

      const xPos = char.offsetLeft - container.offsetLeft;
      const yPos = 10;
      const width = 40;
      const height = 40;

      arrow.setAttribute("x1", `${xPos}`);
      arrow.setAttribute("y1", `${yPos}`);
      arrow.setAttribute("x2", `${xPos}`);
      arrow.setAttribute("y2", `${yPos + height}`);

      rect.setAttribute("x", `${xPos - width / 2}`);
      rect.setAttribute(
        "y",
        `${-(char.offsetTop - container.offsetTop + char.clientHeight + yPos)}`
      );
      rect.setAttribute("width", `${width}`);
      rect.setAttribute("height", `${container.clientHeight}`);
    });
  }

  addChar(addBefore: number) {
    if (this.selected === null || this.charAdded) {
      return;
    }
    this.word.splice(addBefore, 0, {
      id: Math.max(...this.word.map((el) => el.id)) + 1,
      char: alphabet[this.selected],
      initialChar: "",
      locked: false,
    });
    this.selected = null;
    this.charAdded = true;
  }

  undo() {
    this.word = this.word.filter((el) => el.locked);
    this.charAdded = false;
  }

  getAnimationSteps(): Array<string> {
    const [
      correctPos,
      correctChar,
      wrongPos,
      wrongChar,
    ] = findCorrectAndWrongSolutions(
      this.word.map((el) => el.char).join(""),
      this.similarWords[0]
    );

    return new Array<string>(
      `item-selection-${alphabet.findIndex((el) => el === wrongChar)}`,
      `rect-around-arrow-${wrongPos}`,
      "button-menu-check",
      "undo",
      `item-selection-${alphabet.findIndex((el) => el === correctChar)}`,
      `rect-around-arrow-${correctPos}`,
      "button-menu-check",
      "button-menu-next"
    );
  }

  get arrows(): number[] {
    const arrows = new Array<number>();
    for (let i = 0; i <= this.word.length; i++) {
      arrows.push(i);
    }
    return arrows;
  }

  get items(): item[] {
    return items;
  }
}
</script>

<style scoped>
.svg-container {
  width: 100%;
}
.svg-item {
  width: 100%;
  max-height: 3em;
  overflow: visible;
}
</style>
