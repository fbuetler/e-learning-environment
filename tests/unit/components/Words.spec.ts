import { shallowMount, config } from "@vue/test-utils";
import Add from "@/components/words/Add.vue";
import Change from "@/components/words/Change.vue";
import Remove from "@/components/words/Remove.vue";
import Swap from "@/components/words/Swap.vue";
import { wordElement } from "@/components/words/Words";

config.showDeprecationWarnings = false;

function prepareWord(word: string): wordElement[] {
  return word.split("").map((el, i) => {
    return { id: i, char: el, initialChar: el, locked: false };
  });
}

function addChar(
  word: wordElement[],
  char: string,
  pos: number
): wordElement[] {
  word.splice(pos, 0, {
    id: Math.max(...word.map((el) => el.id)) + 1,
    char: char,
    initialChar: "",
    locked: false,
  });
  return word;
}

function changeChar(
  word: wordElement[],
  char: string,
  pos: number
): wordElement[] {
  word.forEach((el, idx) => (el.locked = idx !== pos));
  word[pos].char = char;
  return word;
}

function removeChar(word: wordElement[], pos: number): wordElement[] {
  word[pos].char = "";
  word.forEach((el) => (el.locked = true));
  return word;
}

function swapChar(
  word: wordElement[],
  leftPos: number,
  rightPos: number
): wordElement[] {
  word[leftPos].char = word[rightPos].initialChar;
  word[rightPos].char = word[leftPos].initialChar;
  return word;
}

function copy(el) {
  return JSON.parse(JSON.stringify(el));
}

const rawWord = "ADLER";
const mockWord = prepareWord(rawWord);
const mockSimilarWords = ["RADLER", "TADLER", "ADLERN", "ADLERS"];
const mockFunc = jest.fn();

jest.mock("@/components/words/Words", () => ({
  LoadWords() {
    return [
      JSON.parse(JSON.stringify(mockWord)),
      JSON.stringify(mockSimilarWords).slice(),
    ];
  },
  findCorrectAndWrongSolutions(
    word: string,
    similarWord: string
  ): [number, string, number, string] {
    return [0, "", 0, ""];
  },
  get alphabet() {
    return ["A", "B", "C"];
  },
  get items() {
    return this.alphabet.map((el, i) => {
      return {
        id: i,
        type: i,
        value: i,
        img: el,
        class: "word-char",
      };
    });
  },
}));

describe("Add.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Add, {});
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("mock used correctly", () => {
    expect(wrapper.vm["word"].map((el) => el.char).join("")).toBe(rawWord);
  });

  it("initially all chars are locked and nothing is added", () => {
    expect(wrapper.vm["word"].every((el) => el.locked)).toBeTruthy();
    expect(wrapper.vm["charAdded"]).toBeFalsy();
  });

  for (let i = 0; i <= mockWord.length; i++) {
    it(`adding a char by click at position ${i}`, async () => {
      await wrapper.setData({ selected: 1 }); // 'B'
      wrapper
        .findAll("rect")
        .at(i)
        .trigger("click");
      let newWord = copy(mockWord);
      newWord.forEach((el) => (el.locked = true));
      newWord = addChar(newWord, "B", i);
      expect(wrapper.vm["word"]).toEqual(newWord);
    });
  }

  it("no char selected", () => {
    const newWord = copy(mockWord);
    newWord.forEach((el) => (el.locked = true));
    wrapper.vm.addChar(0);
    expect(copy(wrapper.vm["word"])).toEqual(newWord);
  });

  it("add a char twice", async () => {
    await wrapper.setData({ selected: 1 }); // 'B'
    wrapper.vm.addChar(0);
    const word = copy(wrapper.vm["word"]);
    await wrapper.setData({ selected: 1 }); // 'B'
    wrapper.vm.addChar(0);
    expect(copy(wrapper.vm["word"])).toEqual(word);
  });

  it("correct word is evaluated as correct", async () => {
    await wrapper.setData({ word: prepareWord(mockSimilarWords[0]) });
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect word is evaluated as incorrect", async () => {
    await wrapper.setData({ word: prepareWord("TESTTEST") });
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });

  for (let i = 0; i <= mockWord.length; i++) {
    it(`undo is handled correctly at position ${i}`, async () => {
      await wrapper.setData({ selected: 1 });
      wrapper.find(`#rect-around-arrow-${1}`).trigger("click");
      wrapper.vm.undo();
      const newWord = copy(mockWord);
      newWord.forEach((el) => (el.locked = true));
      expect(wrapper.vm["word"]).toEqual(newWord);
      expect(wrapper.vm["charAdded"]).toBeFalsy();
    });
  }

  it("start next task restores initial conditions", async () => {
    await wrapper.setData({ selected: 1 });
    wrapper.find(`#rect-around-arrow-${1}`).trigger("click");
    await wrapper.setData({ selected: 1 });
    wrapper.vm.restart();
    const newWord = copy(mockWord);
    newWord.forEach((el) => (el.locked = true));
    expect(wrapper.vm["word"]).toEqual(newWord);
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["charAdded"]).toBeFalsy();
  });
});

describe("Change.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Change, {});
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initially all chars are unlocked", () => {
    expect(wrapper.vm["word"].every((el) => !el.locked)).toBeTruthy();
  });

  it("initially nothing is selected", () => {
    expect(wrapper.vm["selected"]).toBeNull();
  });

  for (let i = 0; i < mockWord.length; i++) {
    it(`change char at position ${i}`, async () => {
      await wrapper.setData({ selected: 1 }); // 'B'
      await wrapper.find(`#word-char-${mockWord[i].id}`).trigger("click");
      const newWord = changeChar(copy(mockWord), "B", i);
      expect(wrapper.vm["word"]).toEqual(newWord);
    });
  }

  it("change same char twice", async () => {
    const pos = 0;
    await wrapper.setData({ selected: 1 }); // 'B'
    await wrapper.find(`#word-char-${mockWord[pos].id}`).trigger("click");
    await wrapper.setData({ selected: 2 }); // 'C'
    await wrapper.find(`#word-char-${mockWord[pos].id}`).trigger("click");
    const newWord = changeChar(copy(mockWord), "C", pos);
    expect(wrapper.vm["word"]).toEqual(newWord);
  });

  it("change two different chars", async () => {
    const pos = 0;
    await wrapper.setData({ selected: 1 }); // 'B'
    await wrapper.find(`#word-char-${mockWord[pos].id}`).trigger("click");
    await wrapper.setData({ selected: 2 }); // 'C'
    await wrapper.find(`#word-char-${mockWord[pos + 1].id}`).trigger("click");
    const newWord = changeChar(copy(mockWord), "B", pos);
    expect(wrapper.vm["word"]).toEqual(newWord);
  });

  it("no char selected", async () => {
    await wrapper.setData({ selected: null });
    await wrapper.find(`#word-char-${mockWord[0].id}`).trigger("click");
    expect(wrapper.vm["word"]).toEqual(mockWord);
  });

  for (let i = 0; i < mockWord.length; i++) {
    it(`undo after changing position ${i}`, async () => {
      await wrapper.setData({ selected: 1 }); // 'B'
      await wrapper.find(`#word-char-${mockWord[i].id}`).trigger("click");
      wrapper.vm.undo();
      expect(wrapper.vm["word"]).toEqual(mockWord);
    });
  }

  it("start next task restores initial conditions", async () => {
    await wrapper.setData({ selected: 1 }); // 'B'
    await wrapper.find(`#word-char-${mockWord[0].id}`).trigger("click");
    await wrapper.setData({ selected: 1 }); // 'B'
    wrapper.vm.restart();
    expect(wrapper.vm["word"]).toEqual(mockWord);
    expect(wrapper.vm["selected"]).toBeNull();
  });

  it("correct answer is accepted", async () => {
    await wrapper.setData({ word: prepareWord(mockSimilarWords[0]) });
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect answer is rejected", async () => {
    await wrapper.setData({ word: prepareWord("TESTTEST") });
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });
});

describe("Remove.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Remove, {});
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initially all chars are unlocked", () => {
    expect(wrapper.vm["word"].every((el) => !el.locked)).toBeTruthy();
  });

  for (let i = 0; i < mockWord.length; i++) {
    it(`remove char at position ${i}`, async () => {
      await wrapper.find(`#word-char-${mockWord[i].id}`).trigger("click");
      wrapper.vm.trashElement();
      const newWord = removeChar(copy(mockWord), i);
      expect(wrapper.vm["word"]).toEqual(newWord);
    });
  }

  it("remove two chars", async () => {
    const fstPos = 1;
    const sndPos = 2;
    await wrapper.find(`#word-char-${mockWord[fstPos].id}`).trigger("click");
    wrapper.vm.trashElement();
    await wrapper.find(`#word-char-${mockWord[sndPos].id}`).trigger("click");
    wrapper.vm.trashElement();
    const newWord = removeChar(copy(mockWord), fstPos);
    expect(wrapper.vm["word"]).toEqual(newWord);
  });

  for (let i = 0; i < mockWord.length; i++) {
    it(`undo after removing position ${i}`, async () => {
      await wrapper.find(`#word-char-${mockWord[i].id}`).trigger("click");
      wrapper.vm.trashElement();
      wrapper.vm.undo();
      expect(wrapper.vm["word"]).toEqual(mockWord);
    });
  }

  it("start next task restores initial conditions", async () => {
    await wrapper.find(`#word-char-${mockWord[0].id}`).trigger("click");
    wrapper.vm.trashElement();
    await wrapper.find(`#word-char-${mockWord[1].id}`).trigger("click");
    wrapper.vm.restart();
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["charRemoved"]).toBeFalsy();
    expect(wrapper.vm["word"]).toEqual(mockWord);
  });

  it("correct answer is accepted", async () => {
    await wrapper.setData({ word: prepareWord(mockSimilarWords[0]) });
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect answer is rejected", async () => {
    await wrapper.setData({ word: prepareWord("TESTTEST") });
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });
});

describe("Swap.vue", () => {
  const swapPairs = [
    [0, 1],
    [2, 3],
  ] as [number, number][];
  const mockSwappedWord = swapChar(
    copy(mockWord),
    swapPairs[0][0],
    swapPairs[0][1]
  );
  let wrapper;
  beforeEach(() => {
    const mockRandomNumber = jest
      .fn()
      .mockReturnValueOnce(swapPairs[0][0])
      .mockReturnValueOnce(swapPairs[0][0])
      .mockReturnValueOnce(swapPairs[1][0])
      .mockReturnValue(3);
    wrapper = shallowMount(Swap, {
      methods: {
        randomNumber: mockRandomNumber,
        drawArrows: mockFunc,
        getAnimationSteps: mockFunc,
      },
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initially all chars are unlocked", () => {
    expect(wrapper.vm["word"].every((el) => !el.locked)).toBeTruthy();
  });

  for (let i = 0; i < mockWord.length - 1; i++) {
    it(`swap char at position ${i} with ${i + 1}`, async () => {
      await wrapper.setData({ word: copy(mockWord) });
      await wrapper
        .find(`#rect-around-arrow-${mockWord[i].id}-${mockWord[i + 1].id}`)
        .trigger("click");
      const newWord = swapChar(copy(mockWord), i, i + 1);
      expect(wrapper.vm["word"]).toEqual(newWord);
    });
  }

  it("swap the same char pair", async () => {
    const pos = 2;
    const arrow = wrapper.find(
      `#rect-around-arrow-${mockWord[pos].id}-${mockWord[pos + 1].id}`
    );
    await wrapper.setData({ word: copy(mockWord) });
    await arrow.trigger("click");
    await arrow.trigger("click");
    expect(wrapper.vm["word"]).toEqual(mockWord);
  });

  it("swap two different char pairs", async () => {
    await wrapper.setData({ word: copy(mockWord) });
    await wrapper
      .find(`#rect-around-arrow-${mockWord[0].id}-${mockWord[1].id}`)
      .trigger("click");
    await wrapper
      .find(`#rect-around-arrow-${mockWord[2].id}-${mockWord[3].id}`)
      .trigger("click");
    const newWord = swapChar(swapChar(copy(mockWord), 0, 1), 2, 3);
    expect(wrapper.vm["word"]).toEqual(newWord);
  });

  for (let i = 0; i < mockWord.length - 1; i++) {
    it(`undo after swapping position ${i} with ${i +
      1} restores initial conditions`, async () => {
      await wrapper
        .find(`#rect-around-arrow-${mockWord[i].id}-${mockWord[i + 1].id}`)
        .trigger("click");
      wrapper.vm.undo();
      expect(wrapper.vm["word"]).toEqual(mockSwappedWord);
    });
  }

  it("correct answer is accepted", async () => {
    // TODO (optional): enable after swap has been refactored
    // await wrapper.setData({ word: mockWord });
    // expect(wrapper.vm.isCorrect()).toBeTruthy();
    expect(true).toBeTruthy();
  });

  it("incorrect answer is rejected", async () => {
    // TODO (optional): enable after swap has been refactored
    // await wrapper.setData({ word: prepareWord("TESTTEST") });
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });
});
