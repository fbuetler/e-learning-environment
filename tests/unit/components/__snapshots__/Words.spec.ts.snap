// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add.vue renders correctly 1`] = `
<div>
   
  <div>
    
    Versuche, ein neues Wort zu bilden, indem du einen Buchstaben hinzufügst.
  
  </div>
   
  <div
    class="word-container flex-item flex-col flex-center flex-flex"
    id="word-container"
  >
    <div
      class="flex-item flex-row flex-center"
    >
      <div
        class="word-char card locked"
        id="word-char-0"
      >
        
        A
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-1"
      >
        
        D
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-2"
      >
        
        L
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-3"
      >
        
        E
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-4"
      >
        
        R
      
      </div>
       
      <div />
    </div>
     
    <div
      class="svg-container"
    >
      <svg
        class="svg-item"
      >
        <defs>
          <marker
            id="arrowhead"
            markerHeight="6"
            markerWidth="6"
            orient="auto"
            refX="1"
            refY="5"
            viewBox="0 0 10 10"
          >
            <path
              d="M 0 5 L 10 0 L 10 10 z"
            />
          </marker>
        </defs>
         
        <line
          fill="transparent"
          id="arrow-0"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-1"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-2"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-3"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-4"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-5"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
         
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-0"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-1"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-2"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-3"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-4"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-5"
          width="40"
          x="-20"
          y="-10"
        />
      </svg>
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch flex-wrap"
  >
    <itemselection-stub
      items="[object Object],[object Object],[object Object]"
    />
     
    <undo-stub />
  </div>
</div>
`;

exports[`Change.vue renders correctly 1`] = `
<div>
   
  <div>
    
    Versuche, ein neues Wort zu bilden, indem du einen Buchstaben änderst.
  
  </div>
   
  <div
    class="word-container flex-item flex-row flex-center"
    id="word-container"
  >
    <div>
      <div
        class="word-char card"
        id="word-char-0"
      >
        
        A
      
      </div>
    </div>
    <div>
      <div
        class="word-char card"
        id="word-char-1"
      >
        
        D
      
      </div>
    </div>
    <div>
      <div
        class="word-char card"
        id="word-char-2"
      >
        
        L
      
      </div>
    </div>
    <div>
      <div
        class="word-char card"
        id="word-char-3"
      >
        
        E
      
      </div>
    </div>
    <div>
      <div
        class="word-char card"
        id="word-char-4"
      >
        
        R
      
      </div>
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch flex-wrap"
  >
    <itemselection-stub
      items="[object Object],[object Object],[object Object]"
    />
     
    <undo-stub />
  </div>
</div>
`;

exports[`Remove.vue renders correctly 1`] = `
<div>
   
  <div>
    
    Versuche, ein neues Wort zu bilden, indem du einen Buchstaben entfernst.
  
  </div>
   
  <div
    class="word-container flex-item flex-row flex-center"
    id="word-container"
  >
    <div>
      <div
        class="word-char card clickable"
        draggable="true"
        id="word-char-0"
      >
        
        A
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable"
        draggable="true"
        id="word-char-1"
      >
        
        D
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable"
        draggable="true"
        id="word-char-2"
      >
        
        L
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable"
        draggable="true"
        id="word-char-3"
      >
        
        E
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable"
        draggable="true"
        id="word-char-4"
      >
        
        R
      
      </div>
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch"
  >
    <trashcan-stub />
     
    <undo-stub />
  </div>
</div>
`;

exports[`Swap.vue renders correctly 1`] = `
<div>
   
  <difficulty-stub
    difficultylevels="2"
    selected="1"
  />
   
  <div>
    
    Versuche, ein sinnvolles Wort zu bilden, indem du
    
    <b>
      einmal
    </b>
     
    <!---->
    
    zwei Buchstaben vertauschst.
  
  </div>
   
  <div
    class="word-container flex-item flex-col flex-center flex-flex"
    id="word-container"
  >
    <div
      class="flex-item flex-row flex-center"
    >
      <div
        class="word-char card locked"
        id="word-char-0"
      >
        
        D
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-1"
      >
        
        A
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-2"
      >
        
        L
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-3"
      >
        
        E
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-4"
      >
        
        R
      
      </div>
    </div>
     
    <div
      class="svg-container"
    >
      <svg
        class="svg-item"
      >
        <defs>
          <marker
            id="arrowhead"
            markerHeight="6"
            markerWidth="6"
            orient="auto"
            refX="0"
            refY="5"
            viewBox="0 0 10 10"
          >
            <path
              d="M 10 0 L 0 5 L 10 10 z"
            />
          </marker>
           
          <marker
            id="arrowtail"
            markerHeight="6"
            markerWidth="6"
            orient="auto"
            refX="9"
            refY="5"
            viewBox="0 0 10 10"
          >
            <path
              d="M 0 0 L 10 5 L 0 10 z"
            />
          </marker>
        </defs>
         
        <path
          fill="transparent"
          id="arrow-0-1"
          marker-end="url(#arrowtail)"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="2"
        />
        <path
          fill="transparent"
          id="arrow-1-2"
          marker-end="url(#arrowtail)"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="2"
        />
        <path
          fill="transparent"
          id="arrow-2-3"
          marker-end="url(#arrowtail)"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="2"
        />
        <path
          fill="transparent"
          id="arrow-3-4"
          marker-end="url(#arrowtail)"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="2"
        />
         
        <rect
          class="clickable"
          fill="transparent"
          id="rect-around-arrow-0-1"
        />
        <rect
          class="clickable"
          fill="transparent"
          id="rect-around-arrow-1-2"
        />
        <rect
          class="clickable"
          fill="transparent"
          id="rect-around-arrow-2-3"
        />
        <rect
          class="clickable"
          fill="transparent"
          id="rect-around-arrow-3-4"
        />
      </svg>
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch"
  >
    <undo-stub />
  </div>
</div>
`;
